### ISSUE TITLE
What should be the accessibility of the various LogProvider get logger methods?

### ISSUE THREAD

**Issue by damianh** [damianh/LibLog#48]
In v1-3, `LogProvider.For<>()`, `LogProvider.GetCurrentClassLogger()` and `LogProvider.GetLogger()` were public and available to all.

Reports in the field were of people doing `var logger = YourLib.Logging.LogProvider.For<>()` in _their_ code causing undesirable coupling.

In v4 the intention was to make them internal to prevent this but there is a scenario where this fails: where people have one solution with several projects and they only want to use one LibLog from their `Core` project (and they aren't using `[InternalsVisibleTo]`

Options are:

 1. Make them `public` and add an xml comment something along the lines of "If you don't own this library, it is advisable that you don't calling this."
 2. Make them `internal` only and encourage `[InternalsVisibleTo]` which will have it's own side effects.
 3. ??

**v4 is delisted pending fixing of this**

Ping @SimonCropp @adamralph

**Comment by adamralph** [comment]
I'm still favouring the approach of keeping logging as an internal concern to each assembly. If I put myself in the situation you describe, I would probably just install LibLog in each project in the solution that needs it. The knee jerk reaction from some may be 'but why duplicate all that code?' but the important point is that this is source package and the source itself *is of no concern to the consumer*. It's really no different to installing, say, a binary package Common.Logging into each assembly that needs it. Yes, there's a slight compilation and VCS overhead, but this isn't the 90's.

Given that 99% of assemblies are ultimately going to be consumed in a context that uses one of the LibLog supported concrete logging libs, this will be completely friction free. For the 1% that want to inject their own special logger which can't be expressed via the config of one of the supported logging libs then yes, they have to listen to the set current logger event and propagate and adapt the specified logger to the logger of the inner library, but we are talking about the 1%.

I vote to keep 4.0 as is.

**Comment by adamralph** [comment]
Just realised that that second paragraph probably isn't relevant to the problem at hand.

**Comment by adamralph** [comment]
In fact, there is another option.

In scriptcs, when I installed LiteGuard initially (a much smaller source package) there was opposition to having it installed in every project in the solution. LiteGuard is completely is completely internal by default (it can only be made public by manually hacking the source after installation) so we decided to add the file to only the core project project and then add it as a linked file into all the others. This means a deviation from the default namespace of those other projects, but given that it's all internal this isn't too much of an issue. My personal vote would still be to just install in each project, but this was a reasonable compromise.

**Comment by adamralph** [comment]
I believe the reason I slipped into writing that redundant paragraph in my original comment is because my initial instinct tells me LibLog should be *entirely* internal, but @damianh made a reasonable argument for keeping `ILog`, `ILogProvider`, `LogLevel` and ``LogProvider.SetCurrentLogger()` public. That is, if you want to plug in a logger which can't be expressed by one of the concrete logging libs or if you want to use an old, unserviced lib which uses an old version of LibLog which is no longer able to talk to the logging lib of your choice.

**Comment by damianh** [comment]
As you have identified, the key difference between LiteGuard and LibLog is that LibLog has a _deliberate_ public API aspect. People having to add LibLog to every project is going to complicate things if/when the user needs to supply a custom LogProvider to all of these projects. I don't think people want that cc @simoncropp

Linked file is interesting (I do that for some "utility" system extensions) but that is only something I could document; it's not something I can code into LibLog nor it's nuget package. Of course people can just copy-paste the file in as suggested in the readme. (Aren't we glad it's a single file for these scenarios? :wink: )

So what it is boiling down to is how do I make `ILog`, `LogProvider.SetCurrentLogger()` etc public to consumers of a lib, but only make `LogProvider.For<>()` only available to projects within a solution. I can't see a way to make this possible ootb.

I think the best I think I can do is:

1. Leave `ILog`, `ILogProvider`, `LogLevel`, `LogProvider.SetCurrentLogger()`, `LogProvider.For<>`, `LogProvider.GetLogger()` public by default.
2. Add `LIBLOG_INTERNAL` conditional compilation symbol which will make everything in 1. Internal. Doing so means consumers can't get at `LogProvider` without an `[InternalsVisibleTo]` **Does this go against LibLog philosophy?**
3. Add xml comment note to  `LogProvider.For<>`, `LogProvider.GetLogger()` saying something along the lines of "if you don't own this lib, you probably shouldn't be calling this".  I don't think I can protect against people running-with-scissors.

**Comment by damianh** [comment]
Ping @Erwinvandervalk

**Comment by adamralph** [comment]
Having thought about it a little more, I'm again beginning to question the notion of LibLog having a public API component at all. It's been proposed that there are two use cases for this:

* Plugging in a custom logger without using one of the supported logging providers. Has it been proven that this is required? In what case can I not configure my custom logger using one of the supported providers?

* Adapting an old library which uses an outdated version of LibLog to newer versions of the supported providers. Has this actually been a requirement so far? If not, is there a possibility it could it never happen in practice? I wonder how many people would actually go to the trouble of doing this before pushing for an update of the library to the latest LibLog or switching to an alternate library which hasn't gone stale.

If we can be confident that the supported providers can provide all the logger pluggability we need then there is no need for a public API. You can just configure the logger once and then all the libs will pick it up. LibLog can just be a purely internal abstraction which reaches out into the supported providers for everything it needs.

If I'm missing something obvious, please set me straight :smile:.

**Comment by damianh** [comment]
It was conceived purely as a way to have a super simple logging public API for library developers to _expose to consumers_ of your library. Thus enable them to configure your library to provide logging without you taking dependencies (Common.Logging). Many libs have something like this in place already NEventStore, TopShelf, Raven etc. A variation of the interface made it's way into [katana](https://katanaproject.codeplex.com/SourceControl/latest#src/Microsoft.Owin/Logging/ILogger.cs) and [aspnet 5](https://github.com/aspnet/Logging/blob/dev/src/Microsoft.Framework.Logging.Interfaces/ILogger.cs#L22).

> Plugging in a custom logger without using one of the supported logging providers. Has it been proven that this is required?

Yes, IdSrv have their own TraceProvider afait. 

>  In what case can I not configure my custom logger using one of the supported providers? 

That's the thing, you don't do the configuration, the consumer of your _lib_ does. And they shouldn't be able to get at the the various internal LogProviders  in your Lib at.

> Adapting an old library which uses an outdated version of LibLog to newer versions of the supported providers. Has this actually been a requirement so far? ? If not, is there a possibility it could it never happen in practice?

LibLog hasn't been around long, but thankfully non of the logging fwks have changes their APIs yet. I do know for a fact that the LoupeLogger has a new major version in the works. Will be interesting to see what happens. Anyway, never say never.

> LibLog can just be a purely internal abstraction which reaches out into the supported providers for everything it needs.

The fact that `LibLog` transparently supports logging frameworks transparently is purely :cake: .

Because this "transparent" support is mostly reflection based, it's not going to be the most stable thing in the world in the long term. I personally wouldn't adopt something like this if it didn't have an escape route. (I think supporting EntLib will bite me later, we'll see.).

I also don't support all logging frameworks out there i.e. ETW, MetroLog etc, and never will. 

I _do_ want to support adaptability so that if someone want's to use Common.Logging or whatever in their project, they can adapt `YourLib.Logging.ILogProvider` to it (or to the aspnet5 signature or the katana one or whatever).

Does this make sense? Would be happy to have a f2f over skype.

**Comment by adamralph** [comment]
@damianh thanks, that does make sense and you're putting me straight as I hoped you would. You've clearly spent more time thinking this through than my meagre efforts so far :smile:.

> In what case can I not configure my custom logger using one of the supported providers?
>> That's the thing, you don't do the configuration, the consumer of your lib does. And they shouldn't be able to get at the the various internal LogProviders in your Lib at.

Yes, that's what I meant. When can the consumers not do it? I had a look at the IdSrv host and there's no reason they couldn't configure their custom logger via one of the supported providers. However, I have to admit that adding something NLog just to be able to do that is a lot of unnecessary baggage. I'm sure if you took away the methods and said 'use NLog instead' they wouldn't be very happy campers.

Anyway, thanks for clarifying the philosophy of the lib. I can understand the arguments now and I can also see where you're coming from regarding the 'best-effort' provider support. Indeed, the picture is far wider than the blinkered view I offered above. You've swung my pendulum of thought back and I can see why `ILog`, `ILogProvider`, `LogLevel`, and `LogProvider.SetCurrentLogProvider()` need to be public. Perhaps it's worth wikifying some extracts of this conversation for future ref?

For purpose of discussion, let's call `LogProvider.SetCurrentLogProvider()` the 'Set' API and `LogProvider.For<>()`, `LogProvider.GetCurrentClassLogger()` and `LogProvider.GetLogger()` the 'Get' API.

Set should obviously be public by default and I don't think you should supply an option to internalise it. As you describe, this serves the main purpose of LibLog which is to allow pluggable logging in libs. Unless I'm missing something, it seems that internalising this would be shooting yourself in the foot.

However, I do agree that you're probably going to have to supply a mechanism to switch Get between public and internal. A conditional compilation symbol seems reasonable and better than separate packages. The question is what should be the default. The simple scenario is a solution with one project. The complex scenario is a solution with many projects that want to have one project as the source of loggers. I suggest that Get is internal by default  since, in the simple scenario, it should be internal. It's only required to be public in the complex scenario, so that should be the opt-in. The logical name for the symbol would then be `LIBLOG_PUBLIC` or perhaps something like `LIBLOG_SHARED` would convey the semantics more accurately.

The "don't run with scissors" XML comments can only help. Perhaps they can even link to a wiki entry which explains things in more detail.

**Comment by SimonCropp** [comment]
I understand the philosophy behind what LibLog is trying to solve. And I respect the stance of dont compromise on the design to meet the requirements of people who are doing it wrong. 

> I think supporting EntLib will bite me later, we'll see.

I am not so sure. since entlib changes so rarely it might actually be the easiest one to support.

> A conditional compilation symbol seems reasonable and better than separate packages.

Agreed. 

> The question is what should be the default. I suggest that Get is internal by default since, in the simple scenario, it should be internal. It's only required to be public in the complex scenario, so that should be the opt-in

Agreed

> The "don't run with scissors" XML comments can only help. Perhaps they can even link to a wiki entry which explains things in more detail.

I think a wiki entry is much more valuable here since the nuances of the problem is probably too large to explain in a xml comment

### What broke for Anotar in v4

So in Anotar I added a test for the scenario when one lib logs to liblog where the implementation of liblog is in another assembly. Yes this was explicitly testing the non-recommended approach but at the time since it was a possible use case I wanted to make sure the IL weaving supported it. I now think this is not just a "non-recommended approach" but very hard to achieve in Anotar. The reason is that the Anotar.LibLog packages has a nuget dependency on LibLog. So if you include it in two different projects you will get liblog.cs in both projects. Given the above conversation I think this is the desired result. Someone would need to hack it to instead reference one liblog. 

Anotar also broke since in my integration tests of the liblog ILWeaving I inject a custom logger into liblog and verify the log entries, But this would seem to be a valid scenario, ie it should be trivial for a writer of a library to write tests that can verify the log results.

**Comment by damianh** [comment]
Many thanks for the input lads :)

You have suggested @adamralph I'm going to make the 'Set' API `public`and the 'Get' API `internal` by default with an ability to make it public via compilation symbol.

I'll do the wiki entry for the lib developers and the xml comments for the consumers of the target lib.

@SimonCropp Terribly sorry if this breaks Anotar :|

**Comment by SimonCropp** [comment]
No worries at all. I will change anotar to relect the decisions made here. Thanks for the discussion.

**Comment by SimonCropp** [comment]
Let me know when u push the nuget

**Comment by damianh** [comment]
Sure. Will prob be tomorrow. Since 4.0.0 was a misfire that didn't reach
anyone, any ojections to v4.0.1?

**Comment by SimonCropp** [comment]
+1 for 401 

![6508023065_1b3bf710ce_o](https://cloud.githubusercontent.com/assets/122666/6930634/b64b8370-d850-11e4-9007-e034f2bc6015.jpg)

**Comment by adamralph** [comment]
:+1: 4.0.1 :shipit:

**Comment by damianh** [comment]
@adamralph @SimonCropp 4.0.1 on nuget.org. Cheers.

**Comment by adamralph** [comment]
NIce! Cheers. Will take it for a spin in scriptcs ASAP.

**Comment by adamralph** [comment]
This should be milestone 4.0.1

**Comment by damianh** [comment]
Just noticed, that from a "Set" API perspective, `ILog` doesn't need to be public - it's not exposed via `ILogProvider`. It's there to be able to support the `LogExtensions` which is internal by default.

**Comment by Erwinvandervalk** [comment]
From a usability perspective, making the ILog internal really helps. So yay!

**Comment by adamralph** [comment]
Good catch.

**Comment by damianh** [comment]
@Erwinvandervalk caught it :)

**Comment by damianh** [comment]
LibLog 4.2.0 on nuget.org

**Comment by adamralph** [comment]
Upgraded scriptcs - all good!

**Comment by damianh** [comment]
cool!

**Comment by Erwinvandervalk** [comment]
Same here.. everything is still crunching green
